### ISSUE TITLE
H5::DataSetIException when using HDF5 IOTensor in graph mode

### ISSUE THREAD

**Issue by zaccharieramzi** [tensorflow/io#745]
After #710 was fixed by #732 , I tried it on a single file (which I sent to @yongtang ), and it worked perfectly.

However, when trying it on the whole dataset, I got the following error (I don't know which file is causing this and I don't know how I could know except iterating through all of them):

```
HDF5-DIAG: Error detected in HDF5 (1.10.5) thread 0:
  #000: external/hdf5/src/H5Dio.c line 185 in H5Dread(): could not get a validated dataspace from file_space_id
    major: Invalid arguments to routine
    minor: Bad value
  #001: external/hdf5/src/H5S.c line 254 in H5S_get_validated_dataspace(): selection + offset not within extent
    major: Dataspace
    minor: Out of range
terminate called after throwing an instance of 'H5::DataSetIException'
```

Apparently this error appeared [here](https://github.com/steven-varga/h5cpp/issues/4) in `h5cpp` (which is what is used in tf-io from what I understand) but was fixed. There is not much info though so I don't really know what happened, maybe this is a new case, in which case I probably need to write the issue over there.

If you have no idea what's happening, I can try and identify the one file causing this issue.

**Comment by yongtang** [comment]
@zaccharieramzi I don't have a good idea about the issue, as the related issue in `h5cpp` is not very clear. Though I think I can add some enhancement to the error message of the HDF5 Dataset (PR #748), so that it will shown up the file that triggers the error. 

Once PR #748 is merged, maybe you could give it a try with the nightly, and see which file is causing the issue?

**Comment by yongtang** [comment]
The latest nightly should have the error message enhancement now.

**Comment by zaccharieramzi** [comment]
Hi @yongtang !

Thanks for this update, it's very useful. I actually have a lot of failing files, you can find one here: https://we.tl/t-qJL5uVdB7o.

The failing code is the following (on CPU):
```python
def image_and_kspace_from_h5(fpath):
    spec = {
        '/kspace': tf.TensorSpec(shape=[None, None, None], dtype=tf.complex64),
        '/reconstruction_esc': tf.TensorSpec(shape=[None, None, None], dtype=tf.float32),
    }
    h5_tensors = tfio.IOTensor.from_hdf5(fpath, spec=spec)
    image = h5_tensors('/reconstruction_esc').to_tensor()
    kspace = h5_tensors('/kspace').to_tensor()
    return image, kspace

files_ds = tf.data.Dataset.list_files(f'./*.h5', seed=0)
image_and_kspace_ds = files_ds.map(
    image_and_kspace_from_h5,
    num_parallel_calls=tf.data.experimental.AUTOTUNE,
)
next(iter(image_and_kspace_ds))
next(iter(image_and_kspace_ds))
next(iter(image_and_kspace_ds))
```

Thanks for being so helpful.

**Comment by yongtang** [comment]
@zaccharieramzi With individual files I could run the above code successfully, **tensorflow-io-nightly==0.11.0.dev20200121152445**

I am running the program on Ubuntu 18.04 + python3.

Can you share your environment?

**Comment by zaccharieramzi** [comment]
I just ran with `tensorflow-io-nightly==0.11.0.dev20200121152445` in my python 3.6.8 + ubuntu 16.04 environment. `tensorflow` is 2.1.0.

Did you run with the 2 files I sent you (the one from before and this one) in the directory?

**Comment by yongtang** [comment]
@zaccharieramzi Can you share the previous file again? The file is not available anymore.

**Comment by yongtang** [comment]
@zaccharieramzi By the way, if possible maybe you can removing the line  `num_parallel_calls=tf.data.experimental.AUTOTUNE,` and give it a try again, to see if the issue persists.

HDF5 library by default is not thread safe and requires passing some special flags. I was in the process of trying to make a thread-safe version available but it is not in the current tensorflow-io build yet.

**Comment by zaccharieramzi** [comment]
ok it's exactly that! 
Awesome that you are working on it, can't wait to test it. Should I close this or keep it open?

**Comment by yongtang** [comment]
@zaccharieramzi We can leave the issue open until this is fixed.

The test for thread-safe is in place (skipped now):
https://github.com/tensorflow/io/blob/5a601c02870d0ca37b2ff5b035974db6d22f1bc3/tests/test_io_dataset_eager.py#L898-L904

We will enable the `parallel=2` test case once HDF5 library is compiled with thread-safe.

**Comment by yongtang** [comment]
@zaccharieramzi PR #760 update the hdf5, and added a global mutex to protect thread safety. (hdf5 lib does have the thread-safe mode. However, it is not multi-threaded. Rather it only serialize the API calls. Also the thread-safe mode in hdf5 does not work with C++ APIs. For that I added a global mutex).

Can you give the latest nightly build a try, and see if the issue has been resolved?

**Comment by zaccharieramzi** [comment]
Hi @yongtang !

Just tried with multi-threading again (same code) and tfio `dev20200127123531`, it's still failing with the same error.

I also tried with a fixed number of parallel calls (2), and the error was the same.

**Comment by zaccharieramzi** [comment]
I am not sure how relevant this is, but I wanted to move forward and check that everything was working without multithreading for the IO.
However, when fitting my model, I still get the same error, even when I don't use parallelism. I don't think `fit` is using parallelism without us knowing it (even if `__inference_distributed_function_141` seems to suggest otherwise), so I guess the problem comes from elsewhere:
This is the failing code:

```python
import tensorflow as tf
import tensorflow_io as tfio

def image_and_kspace_from_h5(fpath):
    spec = {
        '/kspace': tf.TensorSpec(shape=[None, None, None], dtype=tf.complex64),
        '/reconstruction_esc': tf.TensorSpec(shape=[None, None, None], dtype=tf.float32),
    }
    h5_tensors = tfio.IOTensor.from_hdf5(fpath, spec=spec)
    image = h5_tensors('/reconstruction_esc').to_tensor()
    image.set_shape((None, 320, 320))
    kspace = h5_tensors('/kspace').to_tensor()
    kspace.set_shape((None, 640, None))
	# I return this fake value just to have an easier model to write
    return image, tf.ones_like(image)

train_path = './'
files_ds = tf.data.Dataset.list_files(f'{train_path}*.h5', seed=0)
image_and_kspace_ds = files_ds.map(
    image_and_kspace_from_h5,
)

im = tf.keras.layers.Input((320, 320))
out = tf.keras.layers.Lambda(lambda x: x)(im)
model = tf.keras.models.Model(im, out)
model.compile(loss='mse', optimizer='adam')

model.fit(image_and_kspace_ds)
```

This will return the following error:
```
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-6-d31d4dab18cb> in <module>
----> 1 model.fit(image_and_kspace_ds)

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    817         max_queue_size=max_queue_size,
    818         workers=workers,
--> 819         use_multiprocessing=use_multiprocessing)
    820 
    821   def evaluate(self,

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in fit(self, model, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    340                 mode=ModeKeys.TRAIN,
    341                 training_context=training_context,
--> 342                 total_epochs=epochs)
    343             cbks.make_logs(model, epoch_logs, training_result, ModeKeys.TRAIN)
    344 

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in run_one_epoch(model, iterator, execution_function, dataset_size, batch_size, strategy, steps_per_epoch, num_samples, mode, training_context, total_epochs)
    126         step=step, mode=mode, size=current_batch_size) as batch_logs:
    127       try:
--> 128         batch_outs = execution_function(iterator)
    129       except (StopIteration, errors.OutOfRangeError):
    130         # TODO(kaftan): File bug about tf function and errors.OutOfRangeError?

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py in execution_function(input_fn)
     96     # `numpy` translates Tensors to values in Eager mode.
     97     return nest.map_structure(_non_none_constant_value,
---> 98                               distributed_function(input_fn))
     99 
    100   return execution_function

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py in __call__(self, *args, **kwds)
    566         xla_context.Exit()
    567     else:
--> 568       result = self._call(*args, **kwds)
    569 
    570     if tracing_count == self._get_tracing_count():

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py in _call(self, *args, **kwds)
    604       # In this case we have not created variables on the first call. So we can
    605       # run the first trace but we should fail if variables are created.
--> 606       results = self._stateful_fn(*args, **kwds)
    607       if self._created_variables:
    608         raise ValueError("Creating variables on a non-first call to a function"

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in __call__(self, *args, **kwargs)
   2361     with self._lock:
   2362       graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
-> 2363     return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
   2364 
   2365   @property

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in _filtered_call(self, args, kwargs)
   1609          if isinstance(t, (ops.Tensor,
   1610                            resource_variable_ops.BaseResourceVariable))),
-> 1611         self.captured_inputs)
   1612 
   1613   def _call_flat(self, args, captured_inputs, cancellation_manager=None):

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in _call_flat(self, args, captured_inputs, cancellation_manager)
   1690       # No tape is watching; skip to running the function.
   1691       return self._build_call_outputs(self._inference_function.call(
-> 1692           ctx, args, cancellation_manager=cancellation_manager))
   1693     forward_backward = self._select_forward_and_backward_functions(
   1694         args,

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in call(self, ctx, args, cancellation_manager)
    543               inputs=args,
    544               attrs=("executor_type", executor_type, "config_proto", config),
--> 545               ctx=ctx)
    546         else:
    547           outputs = execute.execute_with_cancellation(

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     65     else:
     66       message = e.message
---> 67     six.raise_from(core._status_to_exception(e.code, message), None)
     68   except TypeError as e:
     69     keras_symbolic_tensors = [

~/workspace/fastmri-reproducible-benchmark/venv/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

InvalidArgumentError:  unable to process dataset file./file1002534.h5: H5Dread failed
	 [[{{node IO>HDF5ReadableRead}}]]
	 [[IteratorGetNext]] [Op:__inference_distributed_function_141]

Function call stack:
distributed_function
```

And in the logs we can see:
```
HDF5-DIAG: Error detected in HDF5 (1.10.6) thread 0:
  #000: external/hdf5/src/H5Dio.c line 185 in H5Dread(): could not get a validated dataspace from file_space_id
    major: Invalid arguments to routine
    minor: Bad value
  #001: external/hdf5/src/H5S.c line 255 in H5S_get_validated_dataspace(): selection + offset not within extent
    major: Dataspace
    minor: Out of range
2020-01-29 09:53:41.208226: W tensorflow/core/framework/op_kernel.cc:1655] OP_REQUIRES failed at hdf5_kernels.cc:606 : Invalid argument: unable to process dataset file./file1002534.h5: H5Dread failed
2020-01-29 09:53:41.211201: W tensorflow/core/common_runtime/base_collective_executor.cc:217] BaseCollectiveExecutor::StartAbort Invalid argument: unable to process dataset file./file1002534.h5: H5Dread failed
	 [[{{node IO>HDF5ReadableRead}}]]
	 [[IteratorGetNext]]
```

**Comment by zaccharieramzi** [comment]
Hi @yongtang ,

Have you had a chance to reproduce the bug?
Do you know if you or someone else plans to work on this soon? 

Cheers

**Comment by yongtang** [comment]
@zaccharieramzi Sorry for the late reply. I am quite overwhelmed for the past week. I think I might be able to take a look either after Thursday this week or next week.

**Comment by zaccharieramzi** [comment]
Hi @yongtang ,

I hope you are doing well in these difficult times.
I wanted to know if nonetheless you had found time to look at this issue or if you plan on doing so in the near future?

Thanks in advance

**Comment by yongtang** [comment]
@zaccharieramzi Sorry for the delay. I have been mostly focused on audio support for the past month or so. I was hoping to complete the audio support then to move forward for other high priorities ones like hdf5. So far audio support is mostly in place except for one outstanding issue (https://github.com/tensorflow/io/issues/839#issuecomment-604650724). Though I think I pretty much found out a way to get around it so audio will be completed in one or two weeks. Once this is done I will work on hdf5 issue. I noticed there are already several cumulated issues #867 #841 and this one #745, so this is definitely a high priority one.

**Comment by gruossomonica** [comment]
@yongtang and [#791](https://github.com/tensorflow/io/issues/791#issue-566173350)

**Comment by zaccharieramzi** [comment]
As a side note, with all new versions (i.e. tf 2.2 and tfio 0.13), I don't notice this issue anymore but I notice https://github.com/tensorflow/io/issues/841 instead, when training for a large number of steps but without multithreading for the data read.
Btw, I tried multithreading with these new versions and the issue seems to be gone to be replaced by this new #841 .

**Comment by yongtang** [comment]
@zaccharieramzi with PR #922 merged and new tensorflow-io-nightly available, I think this issue is resolved (also see comment in #898).

I will close this issue now but feel free to re-open if issue persists.

**Comment by btranscend** [comment]
im having the same issue/ files check out under h5check and h5dump but when ran, it cant read the data in the files:

```
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: H5Dio.c line 173 in H5Dread(): can't read data
    major: Dataset
    minor: Read failed
  #001: H5Dio.c line 551 in H5D__read(): can't read data
    major: Dataset
    minor: Read failed
  #002: H5Dchunk.c line 1875 in H5D__chunk_read(): unable to read raw data chunk
    major: Low-level I/O
    minor: Read failed
  #003: H5Dchunk.c line 2905 in H5D__chunk_lock(): data pipeline read failed
    major: Data filters
    minor: Filter operation failed
  #004: H5Z.c line 1347 in H5Z_pipeline(): required filter 'deflate' is not registered
    major: Data filters
    minor: Read failed
terminate called after throwing an instance of 'H5::DataSetIException'
*** Aborted at 1611347242 (unix time) try "date -d @1611347242" if you are using GNU date ***
PC: @     0x7f764968efb7 gsignal
```

**Comment by yongtang** [comment]
@btranscend Can you fill a new issue with additional information (e.g., OS distribution, version, python version), and ideally a sample file that triggers the crash?
